### PR TITLE
Project mortar data

### PR DESCRIPTION
**Pull request by kidder**
## Proposed changes

This projects the Boundary History for p-refinement.  This is needed for AMR with LTS.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: tests/Unit/Domain/Structure/Test_OrientationMapHelpers.cpp [hunk 1]
```diff
@@ -567,7 +585,7 @@ void test_2d_slice_with_orientation(const OrientationMap<3>& orientation_map) {
     get(get<ScalarTensor>(expected_vars)) = oriented_mapped_coords[0];
     get<0>(get<Coords<2>>(expected_vars)) = oriented_mapped_coords[0];
     get<1>(get<Coords<2>>(expected_vars)) = oriented_mapped_coords[1];
-    CHECK(oriented_vars == expected_vars);
+    CHECK_VARIABLES_APPROX(oriented_vars, expected_vars);
```
Comment (wthrowe) [inline tests/Unit/Domain/Structure/Test_OrientationMapHelpers.cpp:588]: What's this change about?  Seems unrelated to the rest of the commit.

File: tests/Unit/Evolution/DiscontinuousGalerkin/Test_MortarDataHolder.cpp [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+#include "Framework/TestHelpers.hpp"
+#include "Utilities/GetOutput.hpp"
+#include "Utilities/MakeString.hpp"
+
+namespace evolution::dg {
+namespace {
+template <size_t Dim>
+void test() {
+  MortarDataHolder<Dim> holder{};
+  std::string expected_output = MakeString{} << "Local mortar data:\n"
```
Comment (wthrowe) [inline tests/Unit/Evolution/DiscontinuousGalerkin/Test_MortarDataHolder.cpp]: This could just be a literal instead of using `MakeString`.

File: tests/Unit/Evolution/DiscontinuousGalerkin/Test_MortarDataHolder.cpp [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+#include "Evolution/DiscontinuousGalerkin/MortarDataHolder.hpp"
+#include "Framework/TestHelpers.hpp"
+#include "Utilities/GetOutput.hpp"
+#include "Utilities/MakeString.hpp"
+
+namespace evolution::dg {
+namespace {
+template <size_t Dim>
+void test() {
+  MortarDataHolder<Dim> holder{};
```
Comment (wthrowe) [inline tests/Unit/Evolution/DiscontinuousGalerkin/Test_MortarDataHolder.cpp:38]: [optional] It would be nice if this test tested something beyond an empty state, but since this is adding a test for a previously untested existing class I won't require it.

File: src/Evolution/DiscontinuousGalerkin/Initialization/Mortars.hpp [hunk 4]
```diff
@@ -72,6 +72,96 @@ mortars_apply_impl(const std::vector<std::array<size_t, Dim>>& initial_extents,
...
+        normal_covector_and_magnitude,
+    const gsl::not_null<MortarDataHistoryType*> mortar_data_history,
+    const Mesh<Dim>& new_mesh, const Element<Dim>& new_element,
+    const std::unordered_map<ElementId<Dim>, amr::Info<Dim>>& neighbor_info,
+    const std::pair<Mesh<Dim>, Element<Dim>>& old_mesh_and_element) {
+  const auto& [old_mesh, old_element] = old_mesh_and_element;
+  ASSERT(old_element.id() == new_element.id(),
+         "p-refinement should not have changed the element id");
+
+  std::unordered_set<DirectionalId<Dim>> old_mortar_ids;
```
Comment (wthrowe) [inline src/Evolution/DiscontinuousGalerkin/Initialization/Mortars.hpp]: No need to copy the hash table to another hash table.  The `old_mortar_ids.contains(mortar_id)` check below is the same as `mortar_mesh->contains(mortar_id)`.

File: src/Evolution/DiscontinuousGalerkin/Initialization/Mortars.hpp [hunk 5]
```diff
@@ -72,6 +72,96 @@ mortars_apply_impl(const std::vector<std::array<size_t, Dim>>& initial_extents,
...
+            const auto project_boundary_data =
+                [&new_mortar_mesh, &new_face_mesh, &new_mesh](
+                    const TimeStepId& /* id */,
+                    const gsl::not_null<::evolution::dg::MortarData<Dim>*>
+                        mortar_data) {
+                  p_project(mortar_data, new_mortar_mesh, new_face_mesh,
+                            new_mesh);
+                };
+            local_history.for_each(project_boundary_data);
+            remote_history.for_each(project_boundary_data);
```
Comment (wthrowe) [inline src/Evolution/DiscontinuousGalerkin/Initialization/Mortars.hpp]: This passes the wrong face and volume meshes to `p_project` for the remote side, but I think either the associated remote fields are always unset or the values are never used.  Assuming that's true, add a comment noting that that's why it's OK.

**Review by wthrowe — CHANGES_REQUESTED** [review]

File: tests/Unit/Domain/Structure/Test_OrientationMapHelpers.cpp (hunk 1 above)
Comment (kidder) [reply tests/Unit/Domain/Structure/Test_OrientationMapHelpers.cpp:588]: When I adapted the existing test to also test my new function, it changed the extents of the mesh used to fill oriented_vars which are set based on the logical coords.  This led to a roundoff difference instead of exact equality.

File: tests/Unit/Evolution/DiscontinuousGalerkin/Test_MortarDataHolder.cpp (hunk 3 above)
Comment (kidder) [reply tests/Unit/Evolution/DiscontinuousGalerkin/Test_MortarDataHolder.cpp:38]: ok, updated

File: src/Evolution/DiscontinuousGalerkin/Initialization/Mortars.hpp (hunk 5 above)
Comment (kidder) [reply src/Evolution/DiscontinuousGalerkin/Initialization/Mortars.hpp]: I added a new project function that only projects the mortar data to make this clear.  (I am adding the new function for the event-based refinement PR, so just seemed easier to add it now.)

File: tests/Unit/Domain/Structure/Test_OrientationMapHelpers.cpp (hunk 1 above)
Comment (wthrowe) [reply tests/Unit/Domain/Structure/Test_OrientationMapHelpers.cpp:588]: OK.

**Review by wthrowe — COMMENTED** [review]
You have some warnings and clang-tidy problems, otherwise looks good.  Squash.
